src/api/app/models/token/workflow.rb

```diff
@@ -13,13 +13,55 @@ def call(options)
 
     yaml_file = Workflows::YAMLDownloader.new(scm_webhook.payload, token: self).call
     workflows = Workflows::YAMLToWorkflowsService.new(yaml_file: yaml_file, scm_webhook: scm_webhook, token: self).call
+
+    case
+    when scm_webhook.closed_merged_pull_request?
+      destroy_target_projects(workflows)
+    when scm_webhook.reopened_pull_request?
+      restore_target_projects(workflows)
+    when scm_webhook.new_pull_request?, scm_webhook.updated_pull_request?
+      call_steps(workflows)
+    end
+  rescue Octokit::Unauthorized, Gitlab::Error::Unauthorized => e
+    raise Token::Errors::SCMTokenInvalid, e.message
+  end
+
+  private
+
+  # TODO: This won't work for ConfigureRepositories since it doesn't have target_package and target_project_name
+  def destroy_target_projects(workflows)
+    workflows.each do |workflow|
+      target_packages = workflow.steps.map(&:target_package).uniq
+      delete_subscriptions(target_packages)
+
+      target_project_names = workflow.steps.map(&:target_project_name).uniq
+      Project.where(name: target_project_names).destroy_all
+    end
+  end
+
+  # TODO: This won't work for ConfigureRepositories since it doesn't have target_package and target_project_name
+  def restore_target_projects(workflows)
+    token_user_login = user.login
+
+    workflows.each do |workflow|
+      target_project_names = workflow.steps.map(&:target_project_name).uniq
```

Comment:
I think we should limit this to branch_package steps.

Reply:
We had a TODO about this and it's now done, we ignore `Workflow::Step::ConfigureRepositories` steps in this method. There's nothing to do for them.